FSW/src/util/memUtil.cpp

```diff
@@ -0,0 +1,250 @@
+    digitalWrite(ADC_CHIP_SELECT, LOW);   // set Slave Select pin to low to select chip
+    photodiode16 = SPI.transfer16(0x0000);// transfer data, send 0 to slave, recieve data from ADC
+    digitalWrite(ADC_CHIP_SELECT, HIGH);  // set Slave Select pin to high to de-select chip
+    SPI.endTransaction();
+
+    // Convert ADC output to voltage by scaling over voltage range
+
+    //TODO: can we scale this by the actual digital voltage on the board?
+    voltage = photodiode16 / ADC_BINS * (ADC_MAX_VOLTAGE - ADC_MIN_VOLTAGE);
+
+    return voltage;
+}
+
+/* - - - - - - scienceMemoryHandling - - - - - - *
+ * Usage:
+ *  Takes data from processing and stores it in a buffer,
+ *  continuously overwriting old entries.
+ *  Copies buffer to long term memory when signaled by timing module.
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void scienceMemoryHandling()
+{    
+    if (dataProcessEvent.checkInvoked()){ // checks event status from timing module
+        float currentPhotodiodeVoltage = dataProcessing();
+
+        Serial.print("Current Photodiode Voltage: ");
+        Serial.print(currentPhotodiodeVoltage);
+        Serial.print("\n");
+
+        updateBuffer(currentPhotodiodeVoltage, bufIdx);
+    }
+
+    if (saveEvent.checkInvoked()){
+        saveBuffer(bufIdx);
+    }
+}
+
+/* - - - - - - Helper Functions - - - - - - */
+
+/* - - - - - - updateBuffer - - - - - - *
+ * Belongs to Science Memory Handling Module
+ *  
+ * Usage:
+ *  adds a new sample to the data buffer at specified index
+ *  an existing element at index will be overwritten
+ * 
+ * Inputs:
+ *  dataBuffer - pointer to first element of data buffer
+ *  sample - data to be stored (type must match dataBuffer type)
+ *  index - index to store new data at (pass-by-reference)
+ *  
+ * Outputs:
+ *  none
+ */
+void updateBuffer(float sample, int &index){
+    
+    // check that index is valid
+    if (0 <= index && index < BUFFERSIZE){
+        dataBuffer[index] = sample;
+        index++;
+    
+    } else {
+        Serial.print("WARNING: invalid dataBuffer index ");
+        Serial.print("(Science Memory Handling Module)\n");
+    }
+
+    if (index == BUFFERSIZE){
+        // reset index if we have reached end of buffer
+        index = 0; 
+    }  
+}
+
+/* - - - - - - saveBuffer - - - - - - *
+ * Belongs to Science Memory Handling Module
+ *  
+ * Usage:
+ *  saves the buffer array passed in to a file on the flash module in csv format
+ *  appends timestamp to the end of the file
+ * 
+ * Inputs:
+ *  dataBuffer - pointer to first element of data buffer
+ *  index - index to store new data at (pass-by-reference)
+ *  
+ * Outputs:
+ *  file creation status
+ */
+bool saveBuffer(int &index){
+    
+    // create array to hold data in time ascending order
+    float timeSortArray[BUFFERSIZE];
+    int j = 0; // iterator for sorted array index
+
+    // reorder array so that it is ascending in time
+    for (int i = index; i < BUFFERSIZE; i++){
```

Comment:
I'm a little confused on the utility here. Why would the array of incoming data not already be ascending in time?

Reply:
Secondary: Even if the array was not sorted, would having the timestamp associated with the data point not make this trivial on the ground? Not sure if there's a point to expend resources on sorting here, unless needed for other applications that I am missing.

Reply:
the buffer is continuously overwritten index-by-index and then the entire buffer is saved when a data window ends. We cannot tell when the data window starts, only when it ends on sunset. So at each timestep we overwrite the oldest value in the buffer array which is the length of one data window. When the data window ends, we know that the buffer contains the correct number of data points, but it is not in order. The most recent index overwritten is the location of the most recent value, so we need to reorder the array so that index becomes the end of the array (preserving the relative locations of the other indices). If this is still confusing I can give a better explanation with some diagrams.

To optimize memory usage we are only storing the timestamp of this most recent value and then backsolving for the remaining timestamps on the ground since we know the sampling rate and the data window is quite short.

Reply:
I'm still slightly confused, but I believe you. The backsolving of the timestamps makes sense.